### ISSUE TITLE
"Invalid JSON response from server" when login view is set to a view that is routed to "" or "/"

### ISSUE THREAD

**Issue by mikhalchenko-alexander** [vaadin/flow#19065]
### Description of the bug

I have a security configuration that I took from Vaadin tutorial

```
@Configuration
@EnableWebSecurity
class SecurityConfig: VaadinWebSecurity() {

    override fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http.authorizeHttpRequests { auth ->
            auth.requestMatchers(AntPathRequestMatcher("/public/**"))
                .permitAll()
        }

        setLoginView(http, MainView::class.java)
        return super.filterChain(http)
    }
    // rest of the config
}
```

My `MainView` is a login page, again, from the tutorial

```
@PageTitle("Main")
@Route("/", layout = MainLayout::class)
@AnonymousAllowed
class MainView : VerticalLayout(), BeforeEnterObserver {
    // Login form layout
}
```
When I open any route that is not `@AnonymousAllowed` I redirected to "/", which is correct. However, instead of `MainView` i see an error
```Invalid JSON response from server: [Conent of frontend/index.html]```

If I change `MainView` route to something that is not "" or "/" (e.g. "/login") it works correctly.

### Expected behavior

The "/" route with configuration above should show login form, the `MainView`

### Minimal reproducible example

Full example

```
import com.vaadin.flow.component.ClickEvent
import com.vaadin.flow.component.applayout.AppLayout
import com.vaadin.flow.component.button.Button
import com.vaadin.flow.component.html.H1
import com.vaadin.flow.component.html.Span
import com.vaadin.flow.component.login.LoginForm
import com.vaadin.flow.component.orderedlayout.FlexComponent
import com.vaadin.flow.component.orderedlayout.FlexComponent.JustifyContentMode
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.router.BeforeEnterEvent
import com.vaadin.flow.router.BeforeEnterObserver
import com.vaadin.flow.router.PageTitle
import com.vaadin.flow.router.Route
import com.vaadin.flow.server.auth.AnonymousAllowed
import com.vaadin.flow.spring.security.AuthenticationContext
import com.vaadin.flow.spring.security.VaadinWebSecurity
import jakarta.annotation.security.PermitAll
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.provisioning.InMemoryUserDetailsManager
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.util.matcher.AntPathRequestMatcher


@SpringBootApplication
class VaadinPlaygroundApplication

fun main(args: Array<String>) {
    runApplication<VaadinPlaygroundApplication>(*args)
}

@Configuration
@EnableWebSecurity
class SecurityConfig: VaadinWebSecurity() {

    override fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http.authorizeHttpRequests { auth ->
            auth.requestMatchers(AntPathRequestMatcher("/public/**"))
                .permitAll()
        }

        setLoginView(http, MainView::class.java)
        return super.filterChain(http)
    }

    @Bean
    fun users(): UserDetailsService {
        val user: UserDetails = User.builder()
            .username("user") // password = password with this hash, don't tell anybody :-)
            .password("{bcrypt}$2a$10\$GRLdNijSQMUvl/au9ofL.eDwmoohzzS7.rmNSJZ.0FxO/BTk76klW")
            .roles("USER")
            .build()
        val admin: UserDetails = User.builder()
            .username("admin")
            .password("{bcrypt}$2a$10\$GRLdNijSQMUvl/au9ofL.eDwmoohzzS7.rmNSJZ.0FxO/BTk76klW")
            .roles("USER", "ADMIN")
            .build()
        return InMemoryUserDetailsManager(user, admin)
    }

}

class MainLayout(private val authContext: AuthenticationContext) : AppLayout() {
    init {
        val logo = H1("Vaadin CRM")
        logo.addClassName("logo")
        val header: HorizontalLayout =
            authContext.getAuthenticatedUser(UserDetails::class.java)
                .map { user ->
                    val logout = Button("Logout") { click: ClickEvent<Button?>? -> this.authContext.logout() }
                    val loggedUser = Span("Welcome " + user.getUsername())
                    HorizontalLayout(logo, loggedUser, logout)
                }.orElseGet { HorizontalLayout(logo) }


        // Other page components omitted.
        addToNavbar(header)
    }
}


@PageTitle("Main")
@Route("/", layout = MainLayout::class)
@AnonymousAllowed
class MainView : VerticalLayout(), BeforeEnterObserver {

    private val login = LoginForm()

    init {
        addClassName("login-view")
        setSizeFull()

        alignItems = FlexComponent.Alignment.CENTER
        justifyContentMode = JustifyContentMode.CENTER

        login.action = "login"

        add(H1("Vaadin CRM"), login)
    }

    override fun beforeEnter(beforeEnterEvent: BeforeEnterEvent) {
        // inform the user about an authentication error
        if (beforeEnterEvent.location
                .queryParameters
                .parameters
                .containsKey("error")
        ) {
            login.isError = true
        }
    }

}

@PageTitle("Private page")
@Route("private", layout = MainLayout::class)
@PermitAll
class PrivateView() : HorizontalLayout() {
    private val label: Span = Span("This is private page")

    init {
        add(label)
    }
}

@PageTitle("Public page")
@Route("public", layout = MainLayout::class)
@AnonymousAllowed
class PublicView() : HorizontalLayout() {
    private val label: Span = Span("This is public page")

    init {
        add(label)
    }
}
```

### Versions

- Vaadin / Flow version: 24.3.8
- Java version: 17
- OS version: MacOS 14.3.1
- Browser version (if applicable): Firefox 124.0.1 (64-bit)
- Application Server (if applicable): Tomcat 10.1.19
- IDE (if applicable): IntelliJ IDEA

**Comment by mcollovati** [comment]
With `VaadinWebSecurity` you are not supposed to override the `filterChain` method, but the ` configure(HttpSecurity http) ` method.

If you have seen `filterChain` override in the documentation, please provide the page link, so we can fix it.

Also, `setLoginView()` should be called after `super.configure()`.

```java
    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http.authorizeHttpRequests(auth ->
                    auth.requestMatchers(
                        AntPathRequestMatcher.antMatcher(HttpMethod.GET, "/public/**")).permitAll());  
        super.configure(http);
        setLoginView(http, LoginView.class); 
    }
```

Please try the above configuration and see if it works.

**Comment by mikhalchenko-alexander** [comment]
Thanks @mcollovati . I think that `filterChain` is actually from my experiments, not the tutorial. When log in didn't work for me I tried to use Spring's new approach for security filter chain configuration.

I tried it with `configuration` snippet you provided. Same result, my `MainView` works with "/login", but not with "" or "/" route.

**Comment by mcollovati** [comment]
Could you please provide the complete example project as a GitHub repository or zip file?

**Comment by mikhalchenko-alexander** [comment]
Here https://github.com/mikhalchenko-alexander/vaadin-login-demo

**Comment by mcollovati** [comment]
It looks like that the problem is the blank route on that the login view  (`@Route("")`)
For some reason, this causes Spring Security to forbid Vaadin RPC requests, issuing a redirect to  `/?error` page.
Being XHR requests, the redirect is not followed, and the Flow client fails because it expects JSON content.

If you change the route annotation on the login view to have a path (e.g. `@Route("login")`) the application works.

**Comment by mikhalchenko-alexander** [comment]
Interesting. I added configuration like this
```
override fun configure(web: WebSecurity) {
    web.ignoring().requestMatchers(
        AntPathRequestMatcher.antMatcher(HttpMethod.POST, "/"),
    )
    super.configure(web)
}
```

and it worked. Not sure if it is a proper solution security wise. The original `VaadinWebSecurity.configure(web: WebSecurity)` is no-op. Maybe it makes sense to add similar configuration on the `VaadinWebSecurity` class level as a default config.

**Comment by mcollovati** [comment]
The reason is that if you set "/" as login view, the `UsernamePasswordAuthenticationFilter` will try to authenticate every POST request to "/", including the ones from Vaadin, even if they are permitted through Spring Security.

**Comment by mikhalchenko-alexander** [comment]
I see. To clarify for those who are interested in details. There is a `loginProcessingUrl` in Spring that is responsible for checking user authentication after user is logged in. This URL is `POST "/"` by default. It is handled by a subclass of `AbstractAuthenticationProcessingFilter`. Ignoring its security as I did above is most likely a bad idea. I also tried to change `loginProcessingUrl` to something meaningless (`/12345`) just for experiment. The login page was shown correctly but after that the login completely stopped to work.

So the correct solution to this is probably using another URL for login page, e.g. `/login`. I'm closing this issue.

Thanks @mcollovati for the help ❤️
